cellengine/resources/compensation.py

```diff
@@ -100,36 +101,50 @@ def dataframe_as_html(self):
         """Return the compensation matrix dataframe as HTML."""
         return self.dataframe._repr_html_()
 
-    def apply(self, file: "FcsFile", inplace: bool = False, **kwargs):
-        """
-        Compensate an FcsFile's data.
+    def apply(self, file: FcsFile, inplace=True, **kwargs):
+        """Compensate an FcsFile's data.
 
         Args:
-            file (FcsFile): The FCS file to compensate.
-            inplace (bool): Compensate the file's data in-place.
+            file (FcsFile): The FcsFile to compensate.
+            inplace (bool): If True, modify the `FcsFile.events` with the result.
+                If False, return the compensated events.
             kwargs (Dict):
                 All arguments accepted by `FcsFile.get_events` are accepted here.
         Returns:
             DataFrame: if ``inplace=True``, updates `FcsFile.events` for
                 the target FcsFile
+
+        Returns:
+            DataFrame: The compensated events.
         """
-        data = file.get_events(**kwargs, inplace=True, destination=None)
+
+        if kwargs.items() == file._events_kwargs.items():
+            data = file.events
+        else:
+            data = file.get_events(inplace=inplace, destination=None, **kwargs)
 
         # spill -> comp by inverting
-        inverted = numpy.linalg.inv(self.dataframe)
+        if self._inverted:
+            inverted = self._inverted
+        else:
+            inverted = numpy.linalg.inv(self.dataframe)
+            self._inverted = inverted
 
         # Calculate matrix product for channels matching between file and comp
-        if data and data[self.channels]:
-            comped = data[self.channels]
+        cols = data.columns
+        ix = list(
+            filter(
+                None,
+                [channel if channel in cols else None for channel in self.channels],
+            )
+        )
+        if any(ix):
+            comped = data[ix]
             comped = comped.dot(inverted)  # type: ignore
             comped.columns = self.channels
             data.update(comped)
         else:
             raise IndexError(
                 "No channels from this file match those in the compensation."
             )
-
-        if inplace:
-            file._events = data
-        else:
-            return data
+        return data
```

Comment:
Think this should be reverted, right?

Reply:
It should be; I left it in so you can compare the two implementations. I'll clean it up if we prefer checking kwargs